python_ta/transforms/type_inference_visitor.py

```diff
@@ -224,21 +224,61 @@ def _assign_type(self, target: NodeNG, expr_type: type, node: astroid.Assign) ->
             # Unpacking assignment, e.g. x, y = ...
             if isinstance(expr_type, typing.TupleMeta):
-                elt_inf_types = [self.lookup_inf_type(subtarget, subtarget.name) for subtarget in target.elts]
-                tuple_inf_type = wrap_container(Tuple, *elt_inf_types)
-                return self.type_constraints.unify(tuple_inf_type, expr_type, node)
+                assign_result = self._assign_tuple(target, expr_type, node)
             else:
-                iter_type_result = self._handle_call(target, '__iter__', expr_type)
+                assign_result = self._handle_call(target, '__iter__', expr_type)
                 for subtarget in target.elts:
-                    target_tvar = self.lookup_typevar(subtarget, subtarget.name)
-                    iter_type_result >> (
-                        lambda t: self.type_constraints.unify(target_tvar, t.__args__[0], node)
-                    )
-                return iter_type_result
+                    if isinstance(subtarget, astroid.Starred):
+                        target_tvar = self.lookup_typevar(subtarget.value, subtarget.value.name)
+                        unif_result = assign_result >> (
+                            lambda t: self.type_constraints.unify(target_tvar, List[t.__args__[0]], node))
+                    else:
+                        target_tvar = self.lookup_typevar(subtarget, subtarget.name)
+                        unif_result = assign_result >> (
+                            lambda t: self.type_constraints.unify(target_tvar, t.__args__[0], node))
+
+                    if isinstance(unif_result, TypeFail):
+                        return unif_result
+            return assign_result
         elif isinstance(target, astroid.Subscript):
             # TODO: previous case must recursively handle this one
             return self._handle_call(target, '__setitem__', target.value.inf_type, target.slice.inf_type, expr_type)
 
+    def _assign_tuple(self, target: astroid.Tuple, value: TupleMeta, node: astroid.Assign) -> TypeResult:
+        starred_index = None
+        for i in range(len(target.elts)):
+            if isinstance(target.elts[i], astroid.Starred):
+                if starred_index is None:
+                    starred_index = i
+                else:
+                    return TypeFailStarred(node)
+
+        if starred_index is not None:
+            starred_length = len(value.__args__) - len(target.elts) + 1
+            starred_values = node.value.elts[starred_index:starred_index+starred_length]
+            new_starred_value = wrap_container(List, self.unify_elements(starred_values, node))
+            subvalue_types = list(value.__args__)
+            new_subvalues = subvalue_types[:starred_index] + [new_starred_value] + subvalue_types[starred_index + starred_length:]
+
+            for subtarget, subvalue in zip(target.elts, new_subvalues):
+                if isinstance(subtarget, astroid.Starred):
```

Comment:
Ah, so the fact that you still need to differentiate `Starred` from the other nodes makes this approach not as elegant as the one I suggested. Please switch it.